### PR TITLE
feat: WarpModule with create()

### LINKED ISSUE #3839: Implement `WarpCrudModule` `create()`
Blocked by:
- https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3579

Instead of removing `HypERC20Deployer`, we keep it continue to maintain it with the intent of eventually deprecating

Concretely:
- When creating, just deploy the contracts necessary

### Related CLI Issues
- https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3541
- https://github.com/hyperlane-xyz/issues/issues/1190
- https://github.com/hyperlane-xyz/issues/issues/1156

### PR DESCRIPTION
**Pull request by ltyu**
### Description
- Adds WarpModule with create logic. 
- Updates logic to WarpModuleReader

### Drive by
- Adds CustomIsmConfig mentioned in https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3773

### Related issues
- Fixes #3839 

### Backward compatibility
Yes

### Testing
Unit Tests

### CODE REVIEW

File: typescript/sdk/src/ism/EvmIsmModule.ts [hunk 1]
```diff
@@ -11,79 +11,49 @@ import { MultiProvider } from '../providers/MultiProvider.js';
...
-    this.creator = new EvmIsmCreator(deployer, multiProvider, args.addresses);
   }
 
-  public async read(): Promise<IsmConfig> {
+  public async read(): Promise<DerivedIsmConfigWithAddress> {
     return await this.reader.deriveIsmConfig(this.args.addresses.deployedIsm);
   }
 
-  public async update(config: IsmConfig): Promise<EthersV5Transaction[]> {
+  public async update(_config: IsmConfig): Promise<EthersV5Transaction[]> {
```
Comment (ltyu) [inline typescript/sdk/src/ism/EvmIsmModule.ts:45]: Ignore these IsmModule changes since @paulbalaji PR will implement

Comment (cmcewen) [reply typescript/sdk/src/ism/EvmIsmModule.ts:45]: it seems like you're deleting what will be dead code - so maybe ignore isn't right, but just clarifying these methods will be implemented after a followup PR lands?

File: typescript/sdk/src/ism/EvmIsmReader.ts [hunk 2]
```diff
@@ -79,7 +79,6 @@ export class EvmIsmReader implements IsmReader {
       this.provider,
     );
     const moduleType: ModuleType = await ism.moduleType();
-
```
Comment (cmcewen) [inline typescript/sdk/src/ism/EvmIsmReader.ts:82]: nit: save formatting changes for files you touch as part of the PR

File: typescript/sdk/src/ism/HyperlaneIsmFactory.hardhat-test.ts [hunk 3]
```diff
@@ -338,13 +338,13 @@ describe('HyperlaneIsmFactory', async () => {
...
       const existingIsm = ism.address;
       const domainsBefore = await (ism as DomainRoutingIsm).domains();
-
       // deleting the domain and removing from multiprovider should unenroll the domain
       // NB: we'll deploy new multisigIsms for the domains bc of new factories but the routingIsm address should be the same because of existingIsmAddress
       delete exampleRoutingConfig.domains['test3'];
       multiProvider = multiProvider.intersect([
         TestChainName.test1,
-        'test2',
+        TestChainName.test2,
```
Comment (cmcewen) [inline typescript/sdk/src/ism/HyperlaneIsmFactory.hardhat-test.ts:346]: 👏

File: typescript/sdk/src/ism/types.ts [hunk 4]
```diff
@@ -55,6 +57,7 @@ export function ismTypeToModuleType(ismType: IsmType): ModuleType {
     case IsmType.OP_STACK:
     case IsmType.TEST_ISM:
     case IsmType.PAUSABLE:
+    case IsmType.CUSTOM:
```
Comment (cmcewen) [inline typescript/sdk/src/ism/types.ts:60]: nit: generally prefer alphabetized lists

File: typescript/sdk/src/ism/types.ts [hunk 5]
```diff
@@ -100,8 +103,14 @@ export type TrustedRelayerIsmConfig = {
...
 };
 
+export type CustomIsmConfig = {
+  type: IsmType.CUSTOM;
+  address: Address;
+};
+
 export type IsmConfig =
-  | Address
+  | Address // @todo Remove after https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3773 is implemented
```
Comment (cmcewen) [inline typescript/sdk/src/ism/types.ts:112]: 👏

File: typescript/sdk/src/router/HyperlaneRouterChecker.ts [hunk 6]
```diff
@@ -45,7 +45,7 @@ export class HyperlaneRouterChecker<
     const router = this.app.router(this.app.getContracts(chain));
 
     const checkMailboxClientProperty = async (
-      property: keyof MailboxClientConfig,
+      property: keyof Omit<MailboxClientConfig, 'ismFactoryAddresses'>,
```
Comment (cmcewen) [inline typescript/sdk/src/router/HyperlaneRouterChecker.ts:48]: it looks like the zod schema change with `ismFactoryAddresses` makes it optional but this removes it entirely - is that your intention?

File: typescript/sdk/src/token/EvmERC20WarpRouteReader.ts [hunk 7]
```diff
@@ -44,41 +67,65 @@ export class EvmERC20WarpRouteReader {
...
+    } as DerivedTokenRouterConfig;
+  }
+
+  /**
+   * Derives the token type for a given Warp Route address using specific methods
+   *
+   * @param address - The Warp Route address to derive the token type for.
+   * @returns The derived token type, which can be one of: collateralVault, collateral, native, or synthetic.
+   */
+  async deriveTokenType(address: Address): Promise<DerivedTokenType> {
```
Comment (cmcewen) [inline typescript/sdk/src/token/EvmERC20WarpRouteReader.ts:89]: nit: maybe the argument is called `warpRouteAddress` so that i'm not confused about whether it's the token address or the warp route

File: typescript/sdk/src/token/EvmERC20WarpRouteReader.ts [hunk 8]
```diff
@@ -44,41 +67,65 @@ export class EvmERC20WarpRouteReader {
...
-        await this.evmIsmReader.deriveIsmConfig(
-          fetchedBaseMetadata.interchainSecurityModule,
+    for (const [type, { factory, method }] of Object.entries(contractTypes)) {
+      try {
+        const warpRoute = factory.connect(address, this.provider);
+        await warpRoute[method]();
+        return type as DerivedTokenType;
+      } catch (e) {
+        this.logger.debug(
+          `Error accessing token specific property, implying this is not a ${type} token. Defaulting to ${TokenType.native}.`,
```
Comment (cmcewen) [inline typescript/sdk/src/token/EvmERC20WarpRouteReader.ts:115]: this defaulting comment is only true if you fail for all three methods, right? would suggest adding the log line outside of the for loop if so

File: typescript/sdk/src/token/EvmERC20WarpRouteReader.ts [hunk 9]
```diff
@@ -87,22 +134,37 @@ export class EvmERC20WarpRouteReader {
...
-    return {
+    const metadata: WarpRouteBaseMetadata = {
       mailbox,
       owner,
-      token,
       hook,
-      interchainSecurityModule,
     };
+
+    isZeroishAddress(interchainSecurityModule)
```
Comment (cmcewen) [inline typescript/sdk/src/token/EvmERC20WarpRouteReader.ts:150]: noob question but what does a zero address ISM mean?

File: typescript/sdk/src/ism/EvmIsmModule.ts (hunk 1 above)
Comment (ltyu) [reply typescript/sdk/src/ism/EvmIsmModule.ts:45]: Yeah these files were passed over by @paulbalaji. They're used like interfaces, so I won't be impacted by his PR. 

The idea is that my PR should land first, then his ISM implementations

File: typescript/sdk/src/router/HyperlaneRouterChecker.ts (hunk 6 above)
Comment (ltyu) [reply typescript/sdk/src/router/HyperlaneRouterChecker.ts:48]: Yup, there's a subtle reason for Omit. 

Its because we are using `keyof` which will create a type with all keys, despite it being optional. For this function, we do not want to include `ismFactoryAddresses` at all for `property`

File: typescript/sdk/src/token/EvmERC20WarpRouteReader.ts (hunk 8 above)
Comment (ltyu) [reply typescript/sdk/src/token/EvmERC20WarpRouteReader.ts:115]: This comment should be displayed each time it fails, before the next iteration. The second part is confusing, and I will move it outside. also updated comment

File: typescript/sdk/src/token/EvmERC20WarpRouteReader.ts (hunk 7 above)
Comment (ltyu) [reply typescript/sdk/src/token/EvmERC20WarpRouteReader.ts]: Fixed.

File: typescript/sdk/src/token/EvmERC20WarpRouteReader.ts (hunk 9 above)
Comment (ltyu) [reply typescript/sdk/src/token/EvmERC20WarpRouteReader.ts:157]: good question :)

since we are fetching `interchainSecurityModule` from the warp route contract (in the `promise.all` above), if there are any addresses that are not "set", then solidity by default will return the zero address. in other words, the ISM was not set on the Warp Route contract

File: typescript/sdk/src/hook/types.ts [hunk 10]
```diff
@@ -19,6 +19,7 @@ export enum OnchainHookType {
 }
 
 export enum HookType {
+  ADDRESS = 'address',
```
Comment (paulbalaji) [inline typescript/sdk/src/hook/types.ts:22]: thanks!

File: typescript/sdk/src/ism/HyperlaneIsmFactory.hardhat-test.ts (hunk 3 above)
Comment (paulbalaji) [reply typescript/sdk/src/ism/HyperlaneIsmFactory.hardhat-test.ts:346]: nice, i was trying to use the `TestChainName.XYZ` in my ism update tests too

File: .changeset/smart-dancers-type.md [hunk 11]
```diff
@@ -0,0 +1,5 @@
+---
+'@hyperlane-xyz/sdk': minor
+---
+
+Add EvmWarpCrudModule with create() and update()
```
Comment (paulbalaji) [inline .changeset/smart-dancers-type.md:5]: update with the new class names

File: typescript/sdk/src/token/EvmERC20WarpModule.ts [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+  public async read(): Promise<DerivedTokenRouterConfig> {
+    return this.reader.deriveWarpRouteConfig(
+      this.args.addresses.deployedTokenRoute,
+    );
+  }
+
+  /**
+   * Updates the Warp Route contract with the provided configuration.
+   *
+   * @remark Currently only supports updating ISM or hook
```
Comment (paulbalaji) [inline typescript/sdk/src/token/EvmERC20WarpModule.ts:63]:
```suggestion
   * @remark Currently only supports updating ISM or hook.
```

File: typescript/sdk/src/token/EvmERC20WarpModule.ts [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+   * @returns An array of Ethereum transactions that were executed to update the contract, or an error if the update failed.
+   */
+  public async update(
+    expectedConfig: DerivedTokenRouterConfig,
+  ): Promise<EthersV5Transaction[]> {
+    throw Error('Not implemented');
+  }
+
+  /**
+   * Deploys the Warp Route
```
Comment (paulbalaji) [inline typescript/sdk/src/token/EvmERC20WarpModule.ts:75]:
```suggestion
   * Deploys the Warp Route.
```

File: typescript/sdk/src/token/EvmERC20WarpRouteReader.ts (hunk 9 above)
Comment (paulbalaji) [reply typescript/sdk/src/token/EvmERC20WarpRouteReader.ts:157]: can you leave add that as a comment in code pls, to be able to remember this in the future

File: typescript/sdk/src/token/deploy.hardhat-test.ts [hunk 14]
```diff
@@ -68,38 +94,61 @@ describe('TokenDeployer', async () => {
   });
 
   describe('ERC20WarpRouterReader', async () => {
```
Comment (paulbalaji) [inline typescript/sdk/src/token/deploy.hardhat-test.ts:96]: how annoying would it be to refactor the tests into a file per class

Comment (ltyu) [reply typescript/sdk/src/token/deploy.hardhat-test.ts:96]: Not annoying at all. it was on my todo list :)

File: typescript/sdk/src/token/EvmERC20WarpRouteReader.ts (hunk 9 above)
Comment (yorhodes) [reply typescript/sdk/src/token/EvmERC20WarpRouteReader.ts:158]: ethersjs isnt returning zero address by "default", the address is actually zero in the storage slot onchain
https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/0cf692e7315cb91059a4c0b39c7cdf76151851cc/solidity/contracts/Mailbox.sol#L408 we switch on this in the Mailbox to check whether to use the default ISM or a recipient override

Comment (yorhodes) [reply typescript/sdk/src/token/EvmERC20WarpRouteReader.ts:158]: and I would interpret zeroish address as "undefined" on the ism config

File: typescript/sdk/src/token/EvmERC20WarpRouteReader.ts [hunk 15]
```diff
@@ -111,15 +181,45 @@ export class EvmERC20WarpRouteReader {
...
    * @returns A partial ERC20 metadata object containing the token name, symbol, total supply, and decimals.
    */
-  async fetchTokenMetadata(tokenAddress: Address): Promise<ERC20Metadata> {
-    const erc20 = ERC20__factory.connect(tokenAddress, this.provider);
-    const [name, symbol, totalSupply, decimals] = await Promise.all([
+  async fetchTokenMetadata(
+    type: DerivedTokenType,
+    tokenAddress: Address,
+  ): Promise<ERC20Metadata & { token?: string }> {
+    if (type === TokenType.collateral || type === TokenType.collateralVault) {
```
Comment (yorhodes) [inline typescript/sdk/src/token/EvmERC20WarpRouteReader.ts:188]: can we pull this out into a constant somewhere
"collateral" types

File: typescript/sdk/src/token/EvmERC20WarpRouteReader.ts [hunk 16]
```diff
@@ -111,15 +181,45 @@ export class EvmERC20WarpRouteReader {
...
+      );
+      const token = await erc20.wrappedToken();
+      const { name, symbol, decimals, totalSupply } =
+        await this.fetchERC20Metadata(token);
+
+      return { name, symbol, decimals, totalSupply, token };
+    } else if (type === TokenType.synthetic) {
+      return this.fetchERC20Metadata(tokenAddress);
+    } else {
+      // Assumes Native
```
Comment (yorhodes) [inline typescript/sdk/src/token/EvmERC20WarpRouteReader.ts:201]: I think this is a bad assumption
eg I just added some new token types which would fall under "collateral"

File: typescript/sdk/src/token/EvmERC20WarpRouteReader.ts [hunk 17]
```diff
@@ -1,33 +1,56 @@
...
+export type DerivedTokenType = Extract<
+  TokenType,
+  'collateral' | 'collateralVault' | 'native' | 'synthetic'
 >;
 
-type DerivedERC20WarpRouteConfig = Omit<ERC20RouterConfig, 'type' | 'gas'>;
+export type DerivedTokenRouterConfig = TokenRouterConfig & {
+  type: DerivedTokenType;
+  interchainSecurityModule?: DerivedIsmConfigWithAddress;
+};
```
Comment (yorhodes) [inline typescript/sdk/src/token/EvmERC20WarpRouteReader.ts:41]: not the biggest fan of these subsets/separate types but I think my PR on main already cleaned this up a bit

File: typescript/sdk/src/router/schemas.ts [hunk 18]
```diff
@@ -12,6 +15,7 @@ export const MailboxClientConfigSchema = z.object({
   mailbox: ZHash,
   hook: ZHash.optional(),
   interchainSecurityModule: IsmConfigSchema.optional(),
+  ismFactoryAddresses: ProxyFactoryFactoriesSchema.optional(),
```
Comment (yorhodes) [inline typescript/sdk/src/router/schemas.ts:18]: nice

**Review by yorhodes — CHANGES_REQUESTED** [review]
I think we need zero address to derive to `ism: undefined`
otherwise 
```
module=create(config)
derived = module.read()
assert(derived === config)
```
will throw

File: typescript/sdk/src/token/EvmERC20WarpModule.ts [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+    config: DerivedTokenRouterConfig;
+    multiProvider: MultiProvider;
+  }): Promise<EvmERC20WarpModule> {
+    const { chain, config, multiProvider } = params;
+    const deployer = new HypERC20Deployer(multiProvider);
+    const deployedContracts = (
+      await deployer.deploy({
+        [chain]: config,
+      } as ChainMap<TokenConfig & RouterConfig>)
+    )[chain];
```
[the remark this answers is not in the document]
Comment (ltyu) [reply typescript/sdk/src/token/EvmERC20WarpModule.ts:94]: fixed

File: typescript/sdk/src/token/EvmERC20WarpRouteReader.ts (hunk 17 above)
Comment (ltyu) [reply typescript/sdk/src/token/EvmERC20WarpRouteReader.ts:41]: Removed

File: typescript/sdk/src/token/EvmERC20WarpRouteReader.ts (hunk 15 above)
Comment (ltyu) [reply typescript/sdk/src/token/EvmERC20WarpRouteReader.ts:188]: not sure if this adds much value. 

I believe all of the other collaterals checks are of the variety, config.type === collateral || config.type === collaterlaVault. Only time when the property `type` is checked is in the deployer, which is where i copied this from

File: typescript/sdk/src/token/EvmERC20WarpRouteReader.ts (hunk 9 above)
Comment (ltyu) [reply typescript/sdk/src/token/EvmERC20WarpRouteReader.ts:158]: Fixed.

File: typescript/sdk/src/token/EvmERC20WarpRouteReader.ts (hunk 16 above)
Comment (ltyu) [reply typescript/sdk/src/token/EvmERC20WarpRouteReader.ts]: fixed.

File: typescript/cli/src/config/ism.ts [hunk 20]
```diff
@@ -73,6 +73,7 @@ const ISM_TYPE_DESCRIPTIONS: Record<IsmType, string> = {
     'ISM where you can deliver messages without any validation (WARNING: only for testing, do not use in production)',
   [IsmType.OP_STACK]: '',
   [IsmType.PAUSABLE]: '',
+  [IsmType.CUSTOM]: '',
```
Comment (nbayindirli) [inline typescript/cli/src/config/ism.ts:76]: let's only add these if they're going to be implemented per the recent bug bash

see https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/3908

File: typescript/sdk/src/hook/types.ts [hunk 21]
```diff
@@ -27,6 +27,7 @@ export enum OnchainHookType {
...
```
Comment (yorhodes) [inline typescript/sdk/src/hook/types.ts:30]: can we be consistent with IsmType? either ADDRESS or CUSTOM in both?

File: typescript/sdk/src/token/EvmERC20WarpModule.ts [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+
+import { EvmERC20WarpRouteReader } from './EvmERC20WarpRouteReader.js';
+import { HypERC20Factories } from './contracts.js';
+import { HypERC20Deployer } from './deploy.js';
+import { TokenRouterConfig } from './schemas.js';
+
+export class EvmERC20WarpModule extends HyperlaneModule<
+  ProtocolType.Ethereum,
+  TokenRouterConfig,
+  HyperlaneAddresses<HypERC20Factories> & {
```
Comment (yorhodes) [inline typescript/sdk/src/token/EvmERC20WarpModule.ts:21]: afaik we dont actually serialize with all of these factory keys defined

File: typescript/sdk/src/token/EvmERC20WarpModule.ts [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+    multiProvider: MultiProvider;
+  }): Promise<EvmERC20WarpModule> {
+    const { chain, config, multiProvider } = params;
+    const chainName = multiProvider.getChainName(chain);
+    const deployer = new HypERC20Deployer(multiProvider);
+    const deployedContracts = await deployer.deployContracts(chainName, config);
+
+    return new EvmERC20WarpModule(multiProvider, {
+      addresses: {
+        ...serializeContracts(deployedContracts),
```
Comment (yorhodes) [inline typescript/sdk/src/token/EvmERC20WarpModule.ts:90]: what else might show up here?

File: typescript/sdk/src/token/EvmERC20WarpRouteReader.ts [hunk 24]
```diff
@@ -1,33 +1,42 @@
...
 
-import { TokenType } from './config.js';
-import { TokenRouterConfig } from './schemas.js';
 import { TokenMetadata } from './types.js';
 
-const { AddressZero } = ethers.constants;
+type WarpRouteBaseMetadata = Record<'mailbox' | 'owner', string> & {
+  hook?: HookConfig;
+  interchainSecurityModule?: DerivedIsmConfig;
+};
```
Comment (yorhodes) [inline typescript/sdk/src/token/EvmERC20WarpRouteReader.ts:22]: I am not a fan of "metadata" as a name
further I think this is just `MailboxClient`

File: typescript/sdk/src/token/EvmERC20WarpRouteReader.ts [hunk 25]
```diff
@@ -38,50 +47,89 @@ export class EvmERC20WarpRouteReader {
...
-   * @param address - The address of the Hyperlane ERC20 router contract.
+   * @param warpRouteAddress - The address of the Hyperlane ERC20 router contract.
    * @returns The configuration for the Hyperlane ERC20 router.
    *
    */
   async deriveWarpRouteConfig(
-    address: Address,
-    type = TokenType.collateral,
-  ): Promise<TokenRouterConfig> {
-    const mailboxClientConfig = await this.fetchMailboxClientConfig(address);
```
Comment (yorhodes) [inline typescript/sdk/src/token/EvmERC20WarpRouteReader.ts:49]: revert this change pls

File: typescript/sdk/src/token/EvmERC20WarpRouteReader.ts [hunk 26]
```diff
@@ -112,16 +160,50 @@ export class EvmERC20WarpRouteReader {
...
+        );
+      }
+    } else {
+      throw new Error(
+        `Unsupported token type ${type} when fetching token metadata`,
+      );
+    }
+  }
+
+  async fetchERC20Metadata(tokenAddress: Address): Promise<TokenMetadata> {
```
Comment (yorhodes) [inline typescript/sdk/src/token/EvmERC20WarpRouteReader.ts:198]: I think we have this behavior in a few different places, might be nice to consolidate

**Review by yorhodes — APPROVED** [review]

File: typescript/sdk/src/token/EvmERC20WarpModule.ts (hunk 23 above)
Comment (ltyu) [reply typescript/sdk/src/token/EvmERC20WarpModule.ts:90]: proxyAdmin and timelockController

Comment (ltyu) [reply typescript/sdk/src/token/EvmERC20WarpModule.ts:90]: i can remove them for now

File: typescript/sdk/src/token/EvmERC20WarpModule.ts (hunk 22 above)
Comment (ltyu) [reply typescript/sdk/src/token/EvmERC20WarpModule.ts:21]: removing

File: typescript/sdk/src/token/EvmERC20WarpModule.ts [hunk 27]
```diff
@@ -0,0 +1,106 @@
...
+  });
+  reader: EvmERC20WarpRouteReader;
+
+  constructor(
+    protected readonly multiProvider: MultiProvider,
+    args: HyperlaneModuleArgs<
+      DerivedTokenRouterConfig,
+      HyperlaneAddresses<HypERC20Factories> & {
+        deployedTokenRoute: Address;
+      }
```
[the remark this answers is not in the document]
Comment (ltyu) [reply typescript/sdk/src/token/EvmERC20WarpModule.ts:36]: removed

File: typescript/sdk/src/token/EvmERC20WarpRouteReader.ts (hunk 25 above)
Comment (ltyu) [reply typescript/sdk/src/token/EvmERC20WarpRouteReader.ts:49]: revert to `mailboxClientConfig`

File: typescript/sdk/src/token/EvmERC20WarpRouteReader.ts (hunk 15 above)
Comment (ltyu) [reply typescript/sdk/src/token/EvmERC20WarpRouteReader.ts:169]: moved to `CollateralExtensions`

File: typescript/sdk/src/hook/types.ts (hunk 21 above)
Comment (ltyu) [reply typescript/sdk/src/hook/types.ts]: fixed

File: typescript/cli/src/config/ism.ts (hunk 20 above)
Comment (ltyu) [reply typescript/cli/src/config/ism.ts]: fixed
